util/unifyfs-stage/src/unifyfs-stage.c

```diff
@@ -0,0 +1,292 @@
+    "  -p, --parallel           transfer each file in parallel\n"
+    "                           (experimental)\n"
+    "  -s, --share-dir=<path>   directory path for creating status file\n"
+    "  -v, --verbose            print noisy outputs\n"
+    "\n"
+    "Without the '-p, --parallel' option, a file is transferred by a single\n"
+    "process. If the '-p, --parallel' option is specified, each file will be\n"
+    "divided by multiple processes and transferred in parallel.\n"
+    "\n";
+
+static char* program;
+
+static void print_usage(void)
+{
+    if (0 == rank) {
+        fprintf(stdout, usage_str, program);
+    }
+}
+
+static inline
+void debug_pause(int rank, const char* fmt, ...)
+{
+    if (rank == 0) {
+        va_list args;
+
+        va_start(args, fmt);
+        vfprintf(stderr, fmt, args);
+        va_end(args);
+
+        fprintf(stderr, " ENTER to continue ... ");
+
+        (void) getchar();
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    /* internal accept() call from mpi may set errno */
+    errno = 0;
+}
+
+static int parse_option(int argc, char** argv)
+{
+    int ch = 0;
+    int optidx = 0;
+    char* filepath = NULL;
+
+    if (argc < 2) {
+        return EINVAL;
+    }
+
+    while ((ch = getopt_long(argc, argv,
+                             short_opts, long_opts, &optidx)) >= 0) {
+        switch (ch) {
+        case 'c':
+            checksum = 1;
+            break;
+
+        case 'd':
+            debug = 1;
+            break;
+
+        case 'm':
+            mountpoint = strdup(optarg);
+            break;
+
+        case 'p':
+            mode = UNIFYFS_STAGE_PARALLEL;
+            break;
+
+        case 's':
+            share_dir = strdup(optarg);
+            break;
+
+        case 'v':
+            verbose = 1;
+            break;
+
+        case 'h':
+        default:
+            break;
+        }
+    }
+
+    if (argc - optind != 1) {
+        return EINVAL;
+    }
+
+    filepath = argv[optind];
+
+    manifest_file = realpath(filepath, NULL);
+    if (!manifest_file) {
+        fprintf(stderr, "problem with accessing file %s: %s\n",
+                        filepath, strerror(errno));
+        return errno;
+    }
+
+    return 0;
+}
+
+int main(int argc, char** argv)
+{
+    int ret = 0;
+    unifyfs_stage_t* ctx = &_ctx;
+
+    program = basename(strdup(argv[0]));
+
+    MPI_Init(&argc, &argv);
+    MPI_Comm_size(MPI_COMM_WORLD, &total_ranks);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    ret = parse_option(argc, argv);
+    if (ret) {
+        if (EINVAL == ret) {
+            print_usage();
+        }
+        goto out;
+    }
+
+    ctx->rank = rank;
+    ctx->total_ranks = total_ranks;
+    ctx->checksum = checksum;
+    ctx->mode = mode;
+    ctx->mountpoint = mountpoint;
+    ctx->manifest_file = manifest_file;
+
+    if (verbose) {
+        unifyfs_stage_print(ctx);
+    }
+
+    if (debug) {
+        debug_pause(rank, "About to mount unifyfs.. ");
+    }
```

Comment:
Is this really needed?

Reply:
Don't know.  I'll defer to Hyogi on this.